### PR TITLE
multi parent relationship for changes

### PR DESCRIPTION
**Pull request by samuelstroschein**
Closes https://github.com/opral/lix-sdk/issues/98

**Changes**

- adds `change_edges` table to model multi parent relationships

**Additional information**

- adds TODOS to branch related logic that does not affect the data structure (solving this now is a waste of time if branches changes the behaviour)

### CODE REVIEW

File: lix/packages/sdk/src/change-queue.test.ts [hunk 1]
```diff
@@ -149,46 +145,46 @@ test("should use queue and settled correctly", async () => {
...
 			file_id: "test",
 			plugin_key: "mock-plugin",
 			value: {
-				text: "test",
+				text: "insert text",
 			},
-		},
-		{
+		}),
+		expect.objectContaining({
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/change-queue.test.ts:163]: I updated the tests to `expect.objectContaining` to reduce the refactor effort in the future.

File: lix/packages/sdk/src/database/applySchema.ts [hunk 2]
```diff
@@ -37,7 +36,15 @@ export async function applySchema(args: { sqlite: SqliteDatabase }) {
...
 
-  CREATE INDEX IF NOT EXISTS idx_change_parent_id ON change (parent_id);
+  CREATE TABLE IF NOT EXISTS change_edge (
+    parent_id TEXT NOT NULL,
+    child_id TEXT NOT NULL,
+    PRIMARY KEY (parent_id, child_id),
+    FOREIGN KEY(parent_id) REFERENCES change(id),
+    FOREIGN KEY(child_id) REFERENCES change(id),
+    -- Prevent self referencing edges
+    CHECK (parent_id != child_id)
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/database/applySchema.ts:46]: I didn't know `CHECK` before. Very cool. SQLite now verifies that edges are not self-referencing. A unit test has been written as well.

File: lix/packages/sdk/src/mock/mock-csv-plugin.ts [hunk 3]
```diff
@@ -31,27 +31,13 @@ export const mockCsvPlugin: LixPlugin = {
...
-					.selectFrom("change")
-					.innerJoin("snapshot", "snapshot.id", "change.snapshot_id")
-					.selectAll("change")
-					.select("snapshot.value")
-					.where("change.id", "=", change.parent_id)
-					.executeTakeFirstOrThrow();
-
-				const { rowIndex, columnIndex } = parent.value as unknown as Cell;
-				(parsed.data as any)[rowIndex][columnIndex] = "";
-				// if the row is empty after deleting the cell, remove it
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/mock/mock-csv-plugin.ts:50]: Querying the snapshot is not needed anymore if the entity_id exists on the change itself 😎

File: lix/packages/sdk/src/query-utilities/get-lowest-common-ancestor.ts [hunk 4]
```diff
@@ -12,8 +12,16 @@ export async function getLowestCommonAncestor(args: {
...
 	targetLix: LixReadonly;
 }): Promise<Change | undefined> {
-	// the change has no parent (it is the root change)
-	if (!args.sourceChange?.parent_id) {
+	const parentEdges = await args.sourceLix.db
+		.selectFrom("change_edge")
+		.selectAll()
+		.where("child_id", "=", args.sourceChange.id)
+		// TODO https://github.com/opral/lix-sdk/issues/105
+		// .where(isInSimulatedCurrentBranch)
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/query-utilities/get-lowest-common-ancestor.ts:20]: One of the TODOS that should be tackled after the branch milestone. 

- branching will influence the logic
- working on this now is a waste of time

File: lix/packages/sdk/src/query-utilities/get-lowest-common-ancestor.ts [hunk 5]
```diff
@@ -50,13 +60,26 @@ export async function getLowestCommonAncestor(args: {
...
+			.selectAll()
+			.where("child_id", "=", nextChange.id)
+			.execute();
+		// TODO assumes that only one parent exists
+		if (parentEdges.length > 1) {
+			// https://github.com/opral/inlang-sdk/issues/134
+			throw new Error(
+				"Unimplemented. Multiple parents not supported until the branch feature exists.",
+			);
+		}
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/query-utilities/get-lowest-common-ancestor.ts:81]: One of the TODOS that should be tackled after the branch milestone.

branching will influence the logic
working on this now is a waste of time

File: lix/packages/sdk/src/database/applySchema.ts (hunk 2 above)
Comment (martin-lysk) [reply lix/packages/sdk/src/database/applySchema.ts:46]: I recall we used this to validate the id of the entity type. So if you try to insert an change id (here postfixed with the c) "415f1f62-2017-4761-9ece-519c1e5e45e3-**c**" into a snapshot you can fail early

File: lix/packages/sdk/src/database/initDb.test.ts [hunk 6]
```diff
@@ -73,3 +73,26 @@ test("files should be able to have metadata", async () => {
...
+			.insertInto("change_edge")
+			.values({
+				parent_id: "change1",
+				child_id: "change1",
+			})
+			.returningAll()
+			.execute();
+	} catch (error) {
+		// the sqite3 error is not exposed
+		expect((error as any).name).toBe("SQLite3Error");
```
Comment (martin-lysk) [inline lix/packages/sdk/src/database/initDb.test.ts:112]: Didn't thought of using normal try catch in favour of vitests expect to throw... BUT - this tests passes if the insert doesn't fail? Consider declaring an undefined error before the try, assign the error within the catch and expect after the catch

File: lix/packages/sdk/src/database/applySchema.ts [hunk 7]
```diff
@@ -37,7 +36,15 @@ export async function applySchema(args: { sqlite: SqliteDatabase }) {
...
+  ) strict;
```
Comment (martin-lysk) [inline lix/packages/sdk/src/database/applySchema.ts:47]: We should add an index to the child id as well. I think the parent_id should be covered by the compound primary key - but child_id will most likely need a full table scan in this schema.

**Review by martin-lysk — COMMENTED** [review]
WIP review - will to look deeper into the helper queries

File: lix/packages/sdk/src/database/initDb.test.ts (hunk 6 above)
Comment (samuelstroschein) [reply lix/packages/sdk/src/database/initDb.test.ts:112]: Good one

**Pull request merged**